packages/timeout/src/index.ts

```diff
@@ -0,0 +1,111 @@
+// eslint-disable-next-line max-classes-per-file
+import { setTimeout } from 'timers/promises';
+import { AsyncLocalStorage } from 'async_hooks';
+
+export interface AbortContext {
+    cancelTimeout: AbortController;
+    cancelTask: AbortController;
+}
+
+/**
+ * `AsyncLocalStorage` instance that is used for baring the AbortContext inside user provided handler.
+ * We can use it to access the `AbortContext` instance via `storage.getStore()`, and there we can access
+ * both `cancelTimeout` and `cancelTask` instances of `AbortController`.
+ */
+export const storage = new AsyncLocalStorage<AbortContext>();
+
+/**
+ * Custom error class that will be used for timeout error.
+ */
+export class TimeoutError extends Error {
+}
+
+/**
+ * Custom error class to handle `tryCancel()` checks.
+ * This should not be exposed to user land, as it will be caught in.
+ */
+class InternalTimeoutError extends TimeoutError {
+}
+
+/**
+ * Checks whether we are inside timeout handler created by this package, and cancels current
+ * task execution by throwing `TimeoutError`. This error will be ignored if the promise timed
+ * out already, or explicitly skipped in `addTimeoutToPromise`.
+ *
+ * Use this function after every async call that runs inside the timeout handler:
+ *
+ * ```ts
+ * async function doSomething() {
+ *     await doSomethingTimeConsuming();
+ *     tryCancel();
+ *     await doSomethingElse();
+ *     tryCancel();
+ * }
+ * ```
+ */
+export function tryCancel(): void {
+    const { signal } = storage.getStore()?.cancelTask ?? {};
```

Comment:
Maybe it's just me not being used to new JS features, but it took me a few secs to read it.

```
const signal = storage.getStore()?.cancelTask?.signal; // maybe?
```

Reply:
That's what I got initially, but we have linter that is forcing destructing assignments :D

Reply:
In that case I'd just `// eslint-disable-next-line ...`